rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2160 @@
+
+### Towards an Initial, Smart Implementation
+
+To support transmutations between different types, implementers of this RFC should begin by defining a `transmute_from` lang item to annotate libcore's definition of `TransmuteFrom`. Whether `TransmuteFrom` is implemented for a given type and parameters shall be determined within the implementation of the type system (à la [`Sized`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyS.html#method.is_sized) and [`Freeze`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyS.html#method.is_freeze)).
+
+This initial smart implementation may be made simpler by:
+  - *not* supporting enums
+  - *not* supporting unions
+  - *not* supporting `Neglect` parameters besides `()` and `NeglectStability`
+  - simplifying constructability
+
+#### Simplifying Constructability
+The safety property of constructability defined in the guidance-level examplation of this RFC describes a platonic ideal of the property. 
+
+However, we recognize that this definition poses implementation challenges: In our definition of constructability, answering the question of whether a struct or enum variant is constructible depends on *where* that question is being asked. Consequently, answering whether a given type `Src` is `TransmutableInto` a given type `Dst` will depend on *where* that question is posed.
+
+We recommend adopting a simplified definition of *constructability*: a type is *constructible* if its fields are marked `pub`, and those fields are constructible. With this definition, answering the question of whether a type is constructible does *not* depend on where the question is asked: we do not examine the visibility of the involved types.
+
+Unfortunately, with no other actions taken, this simplified definition comes...
+
+##### ...at the Cost of Safety
+This definition is *usually* sufficient for ensuring safety: it is *generally* an error to expose a private type in a public type signature. However, these errors may be circumvented using the public-type-in-private-module trick:
+```rust
+pub mod crate_a {
+
+    #[repr(C)]
+    pub struct NonEmptySlice<T>(pub private::NonEmptySliceInner<T>);
+
+    impl<T> NonEmptySlice<T> {
+        pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
+            assert!(len <= N && len > 0);
+            Self(
+                private::NonEmptySliceInner {
+                    data: arr as *const T,
+                    len,
+                }
+            )
+        }
+
+        pub fn first(&self) -> &T {
+            unsafe { &*self.0.data }
+        }
+    }
+
+    // introduce a private module to avoid `private_in_public` error (E0446):
+    pub(crate) mod private {
+        #[repr(C)]
+        pub struct NonEmptySliceInner<T> {
+            pub data: *const T,
+            pub len: usize,
+        }
+    }
+
+}
```

Comment:
```suggestion
pub mod crate_a {

    #[repr(C)]
    pub struct NonEmptySlice<'a, T>(pub private::NonEmptySliceInner<'a, T>);

    impl<'a, T> NonEmptySlice<'a, T> {
        pub fn from_array<const N: usize>(arr: &'a [T; N], len: usize) -> Self {
            assert!(len <= N && len > 0);
            Self(
                private::NonEmptySliceInner {
                    data: arr as *const T,
                    len,
                    lifetime: core::marker::PhantomData,
                }
            )
        }

        pub fn first(&self) -> &'a T {
            unsafe { &*self.0.data }
        }
    }

    // introduce a private module to avoid `private_in_public` error (E0446):
    pub(crate) mod private {
        #[repr(C)]
        pub struct NonEmptySliceInner<'a, T> {
            pub data: *const T,
            pub len: usize,
            pub lifetime: core::marker::PhantomData<&'a ()>,
        }
    }

}
```